File: Credit.c
#include <stdio.h>
#include <cs50.h>

long get_number(void);

long get_reverse_number(long card_number);

int main(void)
{
    long card_number = get_number();

    long reverse_card_number = get_reverse_number(card_number);
    int first_digit = (reverse_card_number % 10);
    int second_digit = ((reverse_card_number % 100) / 10);

    int last_digit;

    int i = 0;

    int sum_pairs = 0;
    int sum_odds = 0;
    int sum_total = 0;

    do
    {
    last_digit = (card_number % 10);

    card_number = (card_number / 10);

    if (i % 2 != 0)
    {
        last_digit = last_digit * 2;

        if (last_digit >= 10)
        {
            int memory = 0;

            for (int counter = 0; counter <= 2; counter++)
            {
            memory = last_digit % 10;
            sum_pairs += memory;
            last_digit /= 10;
            }
        }

        else
        {
            sum_pairs += last_digit;
        }
    }

    else
    {
        sum_odds += last_digit;
    }

    i++;

    }
    while (card_number != 0);

    sum_total = (sum_pairs + sum_odds);

    if (sum_total % 10 == 0)
    {
        if (first_digit == 3 && (second_digit == 4 || second_digit == 7) && i == 15)
        {
            printf("AMEX\n");
        }

        else if (first_digit == 4 && (i == 13 || i == 16))
        {
            printf("VISA\n");
        }

        else if (first_digit == 5 && (second_digit == 1 || second_digit == 2 || second_digit == 3 || second_digit == 4 || second_digit == 5) && i == 16)
        {
            printf("MASTERCARD\n");
        }

        else
        {
            printf("INVALID\n");
        }
    }

    else
    {
        printf("INVALID\n");
    }
}

long get_number(void)
{
    long n;

    do
    {
        n = get_long("Number: ");
    }
    while (n <= 0);

    return n;
}

long get_reverse_number(long card_number)
{
   int last_digit_reverse;

   long inverse = 0;

    do
    {
        last_digit_reverse = card_number % 10;
        inverse = (inverse * 10) + last_digit_reverse;
        card_number /= 10;
    }
    while (card_number != 0);

    return inverse;
}

File: Credit.py
def main():

    # Calls the function that gets the card number
    card_number = get_card_number()

    # Converts the card number to a string that will be used later in the code to define the card flag
    card_flag = str(card_number)

    # Some important local variables that will store both sum values
    sum_pairs = 0
    sum_odds = 0

    for i in range(0, len(card_number)):

        last_digit = (int(card_number) % 10)
        card_number = (int(card_number) / 10)

        if (i % 2 != 0):
            last_digit = last_digit * 2

            if last_digit >= 10:
                result = 0

                for digit in str(last_digit):
                    result += int(digit)

                sum_odds += result

            else:
                sum_odds += int(last_digit)

        elif (i % 2 == 0):
            sum_pairs += int(last_digit)

    sum_total = sum_pairs + sum_odds

    # Checks if the last digit of total sum is equal to 0
    if (sum_total % 10) == 0:

        if card_flag[0:2] in ['34', '37'] and (len(card_flag) == 15):
            print('AMEX')

        elif (card_flag[0] == '4') and (len(card_flag) in [14, 16]):
            print('VISA')

        elif (card_flag[0] == '5') and (card_flag[1] in ['1', '2', '3', '4', '5']) and (len(card_flag) == 16):
            print('MASTERCARD')

        else:
            print('INVALID')

    else:
        print('INVALID')


def get_card_number():

    while True:
        n = str(input('Number: '))

        if n.isnumeric():
            if int(n) > 0:
                break
    return n


if __name__ == '__main__':
    main()